crates/bevy_ecs/src/change_detection.rs

```diff
@@ -0,0 +1,152 @@
+    ($name:ident < $( $generics:tt ),+ >, $target:ty, $($traits:ident)?) => {
+        impl<$($generics),* $(: $traits)?> ChangeDetectable for $name<$($generics),*>
+        {
+            #[inline]
+            fn is_added(&self) -> bool {
+                self.ticks
+                    .component_ticks
+                    .is_added(self.ticks.last_change_tick, self.ticks.change_tick)
+            }
+
+            #[inline]
+            fn is_changed(&self) -> bool {
+                self.ticks
+                    .component_ticks
+                    .is_changed(self.ticks.last_change_tick, self.ticks.change_tick)
+            }
+
+            #[inline]
+            fn set_changed(&mut self) {
+                self.ticks
+                    .component_ticks
+                    .set_changed(self.ticks.change_tick);
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> Deref for $name<$($generics),*>
+        {
+            type Target = $target;
+
+            #[inline]
+            fn deref(&self) -> &Self::Target {
+                self.value
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> DerefMut for $name<$($generics),*>
+        {
+            #[inline]
+            fn deref_mut(&mut self) -> &mut Self::Target {
+                self.set_changed();
+                self.value
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> AsRef<$target> for $name<$($generics),*>
+        {
+            #[inline]
+            fn as_ref(&self) -> &$target {
+                self.deref()
+            }
+        }
+
+        impl<$($generics),* $(: $traits)?> AsMut<$target> for $name<$($generics),*>
+        {
+            #[inline]
+            fn as_mut(&mut self) -> &mut $target {
+                self.deref_mut()
+            }
+        }
+    };
+}
+
+macro_rules! impl_into_inner {
+    ($name:ident < $( $generics:tt ),+ >, $target:ty, $($traits:ident)?) => {
+        impl<$($generics),* $(: $traits)?> $name<$($generics),*>
+        {
+            /// Consume `self` and return the contained value while marking `self` as "changed".
+            #[inline]
+            pub fn into_inner(mut self) -> &'a mut $target {
+                self.set_changed();
+                self.value
+            }
+        }
+    };
+}
+
+macro_rules! impl_debug {
+    ($name:ident < $( $generics:tt ),+ >, $($traits:ident)?) => {
+        impl<$($generics),* $(: $traits)?> std::fmt::Debug for $name<$($generics),*>
+            where T: std::fmt::Debug
+        {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                f.debug_tuple(stringify!($name))
+                    .field(self.value)
+                    .finish()
+            }
+        }
+
+    };
+}
+
+pub(crate) struct Ticks<'a> {
+    pub(crate) component_ticks: &'a mut ComponentTicks,
+    pub(crate) last_change_tick: u32,
+    pub(crate) change_tick: u32,
+}
+
+/// Unique borrow of a resource.
```

Comment:
```suggestion
/// Unique mutable borrow of a resource
```

I think this is clearer, and better communicates intent.  The extra period also shouldn't exist.